File: TS/系统/03．技能系统/00．技能模板+函数/04．机制组件/08．机制触发/07．战斗状态触发器.ts
/** @noSelfInFile */

const jass = require("jass.common") as any;

const GetHandleId = jass.GetHandleId as (h: any) => number;
const GetUnitState = jass.GetUnitState as (whichUnit: any, whichUnitState: any) => number;
const IsUnitType = jass.IsUnitType as (whichUnit: any, whichUnitType: any) => boolean;
const UNIT_STATE_MAX_LIFE = jass.UNIT_STATE_MAX_LIFE as any;
const UNIT_TYPE_DEAD = jass.UNIT_TYPE_DEAD as any;

const { registerAppliedFinalDamageListener } = require("系统.04．伤害系统.00．伤害计算.04．主计算流程") as {
  registerAppliedFinalDamageListener: (this: void, cb: (this: void, target: any, attacker: any, applied: number, snapshot: any) => void) => void;
};
const { addPeriodicCallback, removePeriodicCallback, getServerTime } = require("系统.00．核心系统.05．中心计时器") as {
  addPeriodicCallback: (this: void, intervalMs: number, callback: (this: void) => void) => number;
  removePeriodicCallback: (this: void, id: number) => void;
  getServerTime: (this: void) => number;
};
const { 取单位默认脱战时间秒, 取单位默认脱战主体类型, 脱战伤害阈值比例 } = require("系统.00．核心系统.03．脱战系统.00．脱战规则") as {
  取单位默认脱战时间秒: (this: void, unit: any, 指定主体类型?: "玩家英雄" | "Boss" | "普通单位") => number;
  取单位默认脱战主体类型: (this: void, unit: any) => "玩家英雄" | "Boss" | "普通单位";
  脱战伤害阈值比例: number;
};

export interface 战斗状态事件 {
  单位: any;
  对方单位?: any;
  已战斗毫秒: number;
  距离上次战斗毫秒: number;
}

export interface 战斗状态触发参数 {
  名称?: string;
  单位: any;
  主体类型?: "玩家英雄" | "Boss" | "普通单位";
  战斗保持秒?: number;
  检查间隔毫秒?: number;
  周期触发秒?: number;
  持续战斗秒?: number;
  过滤战斗事件?: (this: void, target: any, attacker: any, applied: number, snapshot: any) => boolean;
  on进入战斗?: (this: void, event: 战斗状态事件) => void;
  on脱离战斗?: (this: void, event: 战斗状态事件) => void;
  on周期触发?: (this: void, event: 战斗状态事件) => void;
  on持续战斗满足?: (this: void, event: 战斗状态事件) => void;
}

export interface 战斗状态触发控制器 {
  readonly 名称: string;
  是否战斗中(): boolean;
  刷新战斗(对方单位?: any): void;
  停止(): void;
}

const 战斗状态控制器表: Record<number, 战斗状态触发实现> = {};
let 战斗状态控制器计数 = 0;
let 战斗状态TickID = 0;

function 单位有效(this: void, 单位: any): boolean {
  return 单位 != null && 单位 !== 0 && IsUnitType(单位, UNIT_TYPE_DEAD) !== true;
}

function 取单位ID(this: void, 单位: any): number {
  if (!单位有效(单位)) return 0;
  return GetHandleId(单位) || 0;
}

function 取战斗状态主体类型(this: void, 单位: any, 指定主体类型?: "玩家英雄" | "Boss" | "普通单位"): "玩家英雄" | "Boss" | "普通单位" {
  return 指定主体类型 ?? 取单位默认脱战主体类型(单位);
}

function 受伤达到进入战斗阈值(this: void, 单位: any, 主体类型: "玩家英雄" | "Boss" | "普通单位", applied: number): boolean {
  if (主体类型 !== "玩家英雄") return true;
  const 最大生命 = GetUnitState(单位, UNIT_STATE_MAX_LIFE);
  if (!(最大生命 > 0)) return applied > 0;
  return applied >= 最大生命 * 脱战伤害阈值比例;
}

function 确保战斗状态Tick(this: void, interval: number): void {
  if (战斗状态TickID !== 0) return;
  战斗状态TickID = addPeriodicCallback(interval, on战斗状态Tick);
}

function 尝试停止战斗状态Tick(this: void): void {
  for (const key in 战斗状态控制器表) {
    if (战斗状态控制器表[key] != null) return;
  }
  if (战斗状态TickID !== 0) {
    removePeriodicCallback(战斗状态TickID);
    战斗状态TickID = 0;
  }
}

class 战斗状态触发实现 implements 战斗状态触发控制器 {
  readonly 名称: string;
  readonly 控制器ID: number;
  private 参数: 战斗状态触发参数;
  private 已停止 = false;
  private 战斗中 = false;
  private 战斗开始毫秒 = 0;
  private 上次战斗毫秒 = 0;
  private 下次周期毫秒 = 0;
  private 已触发持续满足 = false;
  private 最近对方单位: any = null;

  constructor(名称: string, 参数: 战斗状态触发参数) {
    this.名称 = 名称;
    this.参数 = 参数;
    this.控制器ID = ++战斗状态控制器计数;
    战斗状态控制器表[this.控制器ID] = this;
    const interval = 参数.检查间隔毫秒 ?? 200;
    确保战斗状态Tick(interval);
  }

  是否战斗中(): boolean {
    return this.战斗中;
  }

  刷新战斗(对方单位?: any): void {
    if (this.已停止 || !单位有效(this.参数.单位)) return;
    const now = getServerTime();
    this.最近对方单位 = 对方单位;
    this.上次战斗毫秒 = now;
    if (!this.战斗中) {
      this.战斗中 = true;
      this.战斗开始毫秒 = now;
      this.下次周期毫秒 = this.取周期触发毫秒(now);
      this.已触发持续满足 = false;
      if (this.参数.on进入战斗 != null) this.参数.on进入战斗(this.创建事件(now));
    }
  }

  处理伤害(target: any, attacker: any, applied: number, snapshot: any): void {
    if (this.已停止 || applied <= 0) return;
    const 单位ID = 取单位ID(this.参数.单位);
    if (单位ID === 0) return;
    const 目标匹配 = 取单位ID(target) === 单位ID;
    const 攻击者匹配 = 取单位ID(attacker) === 单位ID;
    if (!目标匹配 && !攻击者匹配) return;
    if (this.参数.过滤战斗事件 != null && !this.参数.过滤战斗事件(target, attacker, applied, snapshot)) return;
    if (攻击者匹配) {
      this.刷新战斗(target);
      return;
    }
    const 主体类型 = 取战斗状态主体类型(this.参数.单位, this.参数.主体类型);
    if (!受伤达到进入战斗阈值(this.参数.单位, 主体类型, applied)) return;
    this.刷新战斗(attacker);
  }

  Tick(): void {
    if (this.已停止) return;
    if (!单位有效(this.参数.单位)) {
      this.停止();
      return;
    }
    if (!this.战斗中) return;
    const now = getServerTime();
    const 保持秒 = this.参数.战斗保持秒 ?? 取单位默认脱战时间秒(this.参数.单位, this.参数.主体类型);
    const 保持毫秒 = 保持秒 * 1000;
    if (now - this.上次战斗毫秒 > 保持毫秒) {
      const event = this.创建事件(now);
      this.战斗中 = false;
      this.战斗开始毫秒 = 0;
      this.上次战斗毫秒 = 0;
      this.下次周期毫秒 = 0;
      this.已触发持续满足 = false;
      if (this.参数.on脱离战斗 != null) this.参数.on脱离战斗(event);
      return;
    }
    const 持续战斗秒 = this.参数.持续战斗秒 ?? 0;
    if (!this.已触发持续满足 && 持续战斗秒 > 0 && now - this.战斗开始毫秒 >= 持续战斗秒 * 1000) {
      this.已触发持续满足 = true;
      if (this.参数.on持续战斗满足 != null) this.参数.on持续战斗满足(this.创建事件(now));
    }
    if (this.参数.周期触发秒 != null && this.参数.周期触发秒 > 0 && now >= this.下次周期毫秒) {
      if (this.参数.on周期触发 != null) this.参数.on周期触发(this.创建事件(now));
      this.下次周期毫秒 = this.取周期触发毫秒(now);
    }
  }

  停止(): void {
    if (this.已停止) return;
    this.已停止 = true;
    delete 战斗状态控制器表[this.控制器ID];
    尝试停止战斗状态Tick();
  }

  private 创建事件(now: number): 战斗状态事件 {
    return {
      单位: this.参数.单位,
      对方单位: this.最近对方单位,
      已战斗毫秒: this.战斗开始毫秒 > 0 ? now - this.战斗开始毫秒 : 0,
      距离上次战斗毫秒: this.上次战斗毫秒 > 0 ? now - this.上次战斗毫秒 : 0,
    };
  }

  private 取周期触发毫秒(now: number): number {
    const sec = this.参数.周期触发秒 ?? 0;
    return sec > 0 ? now + sec * 1000 : 0;
  }
}

export function 创建战斗状态触发器(this: void, 参数: 战斗状态触发参数): 战斗状态触发控制器 {
  return new 战斗状态触发实现(参数.名称 ?? "战斗状态触发器", 参数);
}

function on战斗状态Tick(this: void): void {
  for (const key in 战斗状态控制器表) {
    const 控制器 = 战斗状态控制器表[key];
    if (控制器 != null) 控制器.Tick();
  }
}

function on战斗状态伤害事件(this: void, target: any, attacker: any, applied: number, snapshot: any): void {
  for (const key in 战斗状态控制器表) {
    const 控制器 = 战斗状态控制器表[key];
    if (控制器 != null) 控制器.处理伤害(target, attacker, applied, snapshot);
  }
}

registerAppliedFinalDamageListener(on战斗状态伤害事件);

File: src/系统/03．技能系统/00．技能模板+函数/04．机制组件/08．机制触发/07．战斗状态触发器.lua
local ____lualib = require("lualib_bundle")
local __TS__Class = ____lualib.__TS__Class
local __TS__Delete = ____lualib.__TS__Delete
local __TS__New = ____lualib.__TS__New
local ____exports = {}
local ____on_6218_6597_72B6_6001Tick, _____6218_6597_72B6_6001_63A7_5236_5668_8868
function ____on_6218_6597_72B6_6001Tick()
    for key in pairs(_____6218_6597_72B6_6001_63A7_5236_5668_8868) do
        local _____63A7_5236_5668 = _____6218_6597_72B6_6001_63A7_5236_5668_8868[key]
        if _____63A7_5236_5668 ~= nil then
            _____63A7_5236_5668:Tick()
        end
    end
end
---
-- @noSelfInFile
local jass = require("jass.common")
local GetHandleId = jass.GetHandleId
local GetUnitState = jass.GetUnitState
local IsUnitType = jass.IsUnitType
local UNIT_STATE_MAX_LIFE = jass.UNIT_STATE_MAX_LIFE
local UNIT_TYPE_DEAD = jass.UNIT_TYPE_DEAD
local ____require_result_0 = require("系统.04．伤害系统.00．伤害计算.04．主计算流程")
local registerAppliedFinalDamageListener = ____require_result_0.registerAppliedFinalDamageListener
local ____require_result_1 = require("系统.00．核心系统.05．中心计时器")
local addPeriodicCallback = ____require_result_1.addPeriodicCallback
local removePeriodicCallback = ____require_result_1.removePeriodicCallback
local getServerTime = ____require_result_1.getServerTime
local ____require_result_2 = require("系统.00．核心系统.03．脱战系统.00．脱战规则")
local _____53D6_5355_4F4D_9ED8_8BA4_8131_6218_65F6_95F4_79D2 = ____require_result_2["取单位默认脱战时间秒"]
local _____53D6_5355_4F4D_9ED8_8BA4_8131_6218_4E3B_4F53_7C7B_578B = ____require_result_2["取单位默认脱战主体类型"]
local _____8131_6218_4F24_5BB3_9608_503C_6BD4_4F8B = ____require_result_2["脱战伤害阈值比例"]
_____6218_6597_72B6_6001_63A7_5236_5668_8868 = {}
local _____6218_6597_72B6_6001_63A7_5236_5668_8BA1_6570 = 0
local _____6218_6597_72B6_6001TickID = 0
local function _____5355_4F4D_6709_6548(_____5355_4F4D)
    return _____5355_4F4D ~= nil and _____5355_4F4D ~= 0 and IsUnitType(_____5355_4F4D, UNIT_TYPE_DEAD) ~= true
end
local function _____53D6_5355_4F4DID(_____5355_4F4D)
    if not _____5355_4F4D_6709_6548(_____5355_4F4D) then
        return 0
    end
    return GetHandleId(_____5355_4F4D) or 0
end
local function _____53D6_6218_6597_72B6_6001_4E3B_4F53_7C7B_578B(_____5355_4F4D, _____6307_5B9A_4E3B_4F53_7C7B_578B)
    return _____6307_5B9A_4E3B_4F53_7C7B_578B or _____53D6_5355_4F4D_9ED8_8BA4_8131_6218_4E3B_4F53_7C7B_578B(_____5355_4F4D)
end
local function _____53D7_4F24_8FBE_5230_8FDB_5165_6218_6597_9608_503C(_____5355_4F4D, _____4E3B_4F53_7C7B_578B, applied)
    if _____4E3B_4F53_7C7B_578B ~= "玩家英雄" then
        return true
    end
    local _____6700_5927_751F_547D = GetUnitState(_____5355_4F4D, UNIT_STATE_MAX_LIFE)
    if not (_____6700_5927_751F_547D > 0) then
        return applied > 0
    end
    return applied >= _____6700_5927_751F_547D * _____8131_6218_4F24_5BB3_9608_503C_6BD4_4F8B
end
local function _____786E_4FDD_6218_6597_72B6_6001Tick(interval)
    if _____6218_6597_72B6_6001TickID ~= 0 then
        return
    end
    _____6218_6597_72B6_6001TickID = addPeriodicCallback(interval, ____on_6218_6597_72B6_6001Tick)
end
local function _____5C1D_8BD5_505C_6B62_6218_6597_72B6_6001Tick()
    for key in pairs(_____6218_6597_72B6_6001_63A7_5236_5668_8868) do
        if _____6218_6597_72B6_6001_63A7_5236_5668_8868[key] ~= nil then
            return
        end
    end
    if _____6218_6597_72B6_6001TickID ~= 0 then
        removePeriodicCallback(_____6218_6597_72B6_6001TickID)
        _____6218_6597_72B6_6001TickID = 0
    end
end
local _____6218_6597_72B6_6001_89E6_53D1_5B9E_73B0 = __TS__Class()
_____6218_6597_72B6_6001_89E6_53D1_5B9E_73B0.name = "战斗状态触发实现"
function _____6218_6597_72B6_6001_89E6_53D1_5B9E_73B0.prototype.____constructor(self, _____540D_79F0, _____53C2_6570)
    self["已停止"] = false
    self["战斗中"] = false
    self["战斗开始毫秒"] = 0
    self["上次战斗毫秒"] = 0
    self["下次周期毫秒"] = 0
    self["已触发持续满足"] = false
    self["最近对方单位"] = nil
    self["名称"] = _____540D_79F0
    self["参数"] = _____53C2_6570
    _____6218_6597_72B6_6001_63A7_5236_5668_8BA1_6570 = _____6218_6597_72B6_6001_63A7_5236_5668_8BA1_6570 + 1
    self["控制器ID"] = _____6218_6597_72B6_6001_63A7_5236_5668_8BA1_6570
    _____6218_6597_72B6_6001_63A7_5236_5668_8868[self["控制器ID"]] = self
    local interval = _____53C2_6570["检查间隔毫秒"] or 200
    _____786E_4FDD_6218_6597_72B6_6001Tick(interval)
end
_____6218_6597_72B6_6001_89E6_53D1_5B9E_73B0.prototype["是否战斗中"] = function(self)
    return self["战斗中"]
end
_____6218_6597_72B6_6001_89E6_53D1_5B9E_73B0.prototype["刷新战斗"] = function(self, _____5BF9_65B9_5355_4F4D)
    if self["已停止"] or not _____5355_4F4D_6709_6548(self["参数"]["单位"]) then
        return
    end
    local now = getServerTime()
    self["最近对方单位"] = _____5BF9_65B9_5355_4F4D
    self["上次战斗毫秒"] = now
    if not self["战斗中"] then
        self["战斗中"] = true
        self["战斗开始毫秒"] = now
        self["下次周期毫秒"] = self["取周期触发毫秒"](self, now)
        self["已触发持续满足"] = false
        if self["参数"]["on进入战斗"] ~= nil then
            self["参数"]["on进入战斗"](self["创建事件"](self, now))
        end
    end
end
_____6218_6597_72B6_6001_89E6_53D1_5B9E_73B0.prototype["处理伤害"] = function(self, target, attacker, applied, snapshot)
    if self["已停止"] or applied <= 0 then
        return
    end
    local _____5355_4F4DID = _____53D6_5355_4F4DID(self["参数"]["单位"])
    if _____5355_4F4DID == 0 then
        return
    end
    local _____76EE_6807_5339_914D = _____53D6_5355_4F4DID(target) == _____5355_4F4DID
    local _____653B_51FB_8005_5339_914D = _____53D6_5355_4F4DID(attacker) == _____5355_4F4DID
    if not _____76EE_6807_5339_914D and not _____653B_51FB_8005_5339_914D then
        return
    end
    if self["参数"]["过滤战斗事件"] ~= nil and not self["参数"]["过滤战斗事件"](target, attacker, applied, snapshot) then
        return
    end
    if _____653B_51FB_8005_5339_914D then
        self["刷新战斗"](self, target)
        return
    end
    local _____4E3B_4F53_7C7B_578B = _____53D6_6218_6597_72B6_6001_4E3B_4F53_7C7B_578B(self["参数"]["单位"], self["参数"]["主体类型"])
    if not _____53D7_4F24_8FBE_5230_8FDB_5165_6218_6597_9608_503C(self["参数"]["单位"], _____4E3B_4F53_7C7B_578B, applied) then
        return
    end
    self["刷新战斗"](self, attacker)
end
function _____6218_6597_72B6_6001_89E6_53D1_5B9E_73B0.prototype.Tick(self)
    if self["已停止"] then
        return
    end
    if not _____5355_4F4D_6709_6548(self["参数"]["单位"]) then
        self["停止"](self)
        return
    end
    if not self["战斗中"] then
        return
    end
    local now = getServerTime()
    local _____4FDD_6301_79D2 = self["参数"]["战斗保持秒"] or _____53D6_5355_4F4D_9ED8_8BA4_8131_6218_65F6_95F4_79D2(self["参数"]["单位"], self["参数"]["主体类型"])
    local _____4FDD_6301_6BEB_79D2 = _____4FDD_6301_79D2 * 1000
    if now - self["上次战斗毫秒"] > _____4FDD_6301_6BEB_79D2 then
        local event = self["创建事件"](self, now)
        self["战斗中"] = false
        self["战斗开始毫秒"] = 0
        self["上次战斗毫秒"] = 0
        self["下次周期毫秒"] = 0
        self["已触发持续满足"] = false
        if self["参数"]["on脱离战斗"] ~= nil then
            self["参数"]["on脱离战斗"](event)
        end
        return
    end
    local _____6301_7EED_6218_6597_79D2 = self["参数"]["持续战斗秒"] or 0
    if not self["已触发持续满足"] and _____6301_7EED_6218_6597_79D2 > 0 and now - self["战斗开始毫秒"] >= _____6301_7EED_6218_6597_79D2 * 1000 then
        self["已触发持续满足"] = true
        if self["参数"]["on持续战斗满足"] ~= nil then
            self["参数"]["on持续战斗满足"](self["创建事件"](self, now))
        end
    end
    if self["参数"]["周期触发秒"] ~= nil and self["参数"]["周期触发秒"] > 0 and now >= self["下次周期毫秒"] then
        if self["参数"]["on周期触发"] ~= nil then
            self["参数"]["on周期触发"](self["创建事件"](self, now))
        end
        self["下次周期毫秒"] = self["取周期触发毫秒"](self, now)
    end
end
_____6218_6597_72B6_6001_89E6_53D1_5B9E_73B0.prototype["停止"] = function(self)
    if self["已停止"] then
        return
    end
    self["已停止"] = true
    __TS__Delete(_____6218_6597_72B6_6001_63A7_5236_5668_8868, self["控制器ID"])
    _____5C1D_8BD5_505C_6B62_6218_6597_72B6_6001Tick()
end
_____6218_6597_72B6_6001_89E6_53D1_5B9E_73B0.prototype["创建事件"] = function(self, now)
    return {["单位"] = self["参数"]["单位"], ["对方单位"] = self["最近对方单位"], ["已战斗毫秒"] = self["战斗开始毫秒"] > 0 and now - self["战斗开始毫秒"] or 0, ["距离上次战斗毫秒"] = self["上次战斗毫秒"] > 0 and now - self["上次战斗毫秒"] or 0}
end
_____6218_6597_72B6_6001_89E6_53D1_5B9E_73B0.prototype["取周期触发毫秒"] = function(self, now)
    local sec = self["参数"]["周期触发秒"] or 0
    return sec > 0 and now + sec * 1000 or 0
end
____exports["创建战斗状态触发器"] = function(_____53C2_6570)
    return __TS__New(_____6218_6597_72B6_6001_89E6_53D1_5B9E_73B0, _____53C2_6570["名称"] or "战斗状态触发器", _____53C2_6570)
end
local function ____on_6218_6597_72B6_6001_4F24_5BB3_4E8B_4EF6(target, attacker, applied, snapshot)
    for key in pairs(_____6218_6597_72B6_6001_63A7_5236_5668_8868) do
        local _____63A7_5236_5668 = _____6218_6597_72B6_6001_63A7_5236_5668_8868[key]
        if _____63A7_5236_5668 ~= nil then
            _____63A7_5236_5668["处理伤害"](
                _____63A7_5236_5668,
                target,
                attacker,
                applied,
                snapshot
            )
        end
    end
end
registerAppliedFinalDamageListener(____on_6218_6597_72B6_6001_4F24_5BB3_4E8B_4EF6)
return ____exports
